src/dali/abstract_ccxt_pair_converter_plugin.py

```diff
@@ -0,0 +1,1027 @@
+
+        # Sort the optimizations by timestamp
+        # while caching the graph snapshots, the partial optimizations will be layered on to the previous
+        # timestamp's optimizations
+        sorted_optimizations = dict(sorted(optimizations.items(), key=lambda x: x[0]))
+
+        # Copy over assets from previous timestamps so there are no holes in the graph
+        composite_optimizations: Dict[datetime, Dict[str, Dict[str, float]]] = {}
+        previous_timestamp: Optional[datetime] = None
+
+        for timestamp, optimization_dict in sorted_optimizations.items():
+            if previous_timestamp is None:
+                composite_optimizations[timestamp] = optimization_dict
+            else:
+                composite_optimizations[timestamp] = sorted_optimizations[previous_timestamp].copy()
+                composite_optimizations[timestamp].update(sorted_optimizations[timestamp])
+
+        previous_assets: Optional[Dict[str, Dict[str, float]]] = None
+        timestamps_to_delete: List[datetime] = []
+        # Next, we assign weights based on the rank of the volume
+        for timestamp, snapshot_assets in composite_optimizations.items():
+            for asset, neighbors in snapshot_assets.items():
+                ranked_neighbors: Dict[str, float] = dict(sorted(iter(neighbors.items()), key=lambda x: x[1], reverse=True))
+                weight: float = 1.0
+                for neighbor_name, neighbor_volume in ranked_neighbors.items():
+                    if neighbor_volume != -1.0:
+                        neighbors[neighbor_name] = weight
+                        weight += 1.0
+                    else:
+                        neighbors[neighbor_name] = -1.0
+
+            # mark duplicate successive snapshots
+            if snapshot_assets == previous_assets:
+                timestamps_to_delete.append(timestamp)
+
+            previous_assets = snapshot_assets
+
+        # delete duplicates
+        for timestamp in timestamps_to_delete:
+            del composite_optimizations[timestamp]
+
+        return composite_optimizations
+
+    def _process_aliases(self, aliases: str) -> Dict[str, Dict[Alias, RP2Decimal]]:
+        alias_list: List[str] = aliases.split(";")
+        processed_aliases: Dict[str, Dict[Alias, RP2Decimal]] = {}
+
+        for alias in alias_list:
+            alias_properties: List[str] = alias.split(",")
+            if alias_properties[0] in _EXCHANGE_DICT or alias_properties[0] == _UNIVERSAL:
+                exchange: str = alias_properties[0]
+            else:
+                raise RP2ValueError(f"Exchange - {alias_properties[0]} is not supported at this time. Check the spelling of the exchange.")
+
+            processed_aliases.setdefault(exchange, {}).update({Alias(alias_properties[1], alias_properties[2]): RP2Decimal(alias_properties[3])})
+
+        return processed_aliases
+
+    ### Abstract Fiat Methods (Must be overridden) ###
```

Comment:
I would just remove this comment: some of the methods below are not abstract, and the abstract one are clear from the raise clause in them.